[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 opened the pull request:
Resolves #7341

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

fashxp commented:
@mcop1  do we need to provide a BC layer in Pimcore 10.5 for that ... especially in API calls where javascript might be depending on `o_id` vs. `id` in the payload.
For example like the get object data call [here](https://github.com/pimcore/pimcore/pull/13308/files#diff-223386524f4ff18dc1e37479f7e783a5e463cce8417be7519db1a05185549f1aL825)?

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

# Conflicts:
#	.github/ci/files/var/classes/DataObject/FilterDefinition.php
#	.github/ci/files/var/classes/DataObject/OnlineShopOrder.php
#	.github/ci/files/var/classes/DataObject/OnlineShopOrderItem.php
#	.github/ci/files/var/classes/DataObject/OnlineShopTaxClass.php
#	.github/ci/files/var/classes/DataObject/OnlineShopVoucherSeries.php
#	.github/ci/files/var/classes/DataObject/OnlineShopVoucherToken.php
#	bundles/AdminBundle/public/js/pimcore/object/folder.js
#	bundles/AdminBundle/public/js/pimcore/object/object.js
#	bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php
#	bundles/AdminBundle/src/Controller/Admin/TagsController.php
#	bundles/AdminBundle/src/Helper/QueryParams.php
#	bundles/EcommerceFrameworkBundle/src/Controller/AdminOrderController.php
#	bundles/EcommerceFrameworkBundle/src/IndexService/Config/ElasticSearch.php
#	bundles/EcommerceFrameworkBundle/src/IndexService/Worker/DefaultFindologic.php
#	bundles/EcommerceFrameworkBundle/src/IndexService/Worker/OptimizedMysql.php
#	doc/Development_Documentation/05_Objects/01_Object_Classes/05_Class_Settings/55_Preview.md
#	doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md
#	models/DataObject/AbstractObject.php
#	models/DataObject/Concrete.php
#	models/DataObject/Folder.php
#	models/DataObject/Objectbrick/Data/Dao.php
#	models/DataObject/Service.php
#	models/Element/Traits/DirtyIndicatorTrait.php
#	tests/Support/Helper/AbstractTestDataHelper.php

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

# Conflicts:
#	models/DataObject/Concrete.php

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Merged 11.x and added method for version dependent field names

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Added method for version dependent field names

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Removed getter function for version dependent field names, since it has to be implemented in 10.6 now
Substituted one more occurrence of o_id

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

mcop1 commented on models/DataObject/ClassDefinition/Dao.php:
```diff
@@ -114,15 +114,15 @@ public function update()
 
         // create object table if not exists
-        $protectedColumns = ['oo_id', 'oo_classId', 'oo_className'];
+        $protectedColumns = ['oo_id', 'oclassId', 'oo_className'];
```
```suggestion
        $protectedColumns = ['oo_id', 'oo_classId', 'oo_className'];
```

mcop1 committed:
Update models/DataObject/ClassDefinition/Dao.php

aashan10 commented on bundles/CoreBundle/src/Migrations/Version20221003115124.php:
```diff
@@ -0,0 +1,95 @@
+
+    public function up(Schema $schema): void
+    {
+        $this->addSql("ALTER TABLE pimcore.objects CHANGE o_id id int(11) unsigned auto_increment NOT NULL;");
```
Should we be adding the table name like `pimcore.object`? Isn't the database name configurable?

brusch replied:
DB-name can be just removed as not needed.

mcop1 replied:
I´ll remove the database name

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
removed database name from migration

mcop1 commented:
> @mcop1 do we need to provide a BC layer in Pimcore 10.5 for that ... especially in API calls where javascript might be depending on `o_id` vs. `id` in the payload. For example like the get object data call [here](https://github.com/pimcore/pimcore/pull/13308/files#diff-223386524f4ff18dc1e37479f7e783a5e463cce8417be7519db1a05185549f1aL825)?

Just for completeness: 
PHP: There is a bc method which returns the version dependent db column name: https://github.com/pimcore/pimcore/issues/13512
JS: We can use something like `foo['o_id'] ?? foo['id']` as we did with childs/children.

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

fashxp commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1366,22 +1366,22 @@ protected function updateIndexesOfObjectSiblings(DataObject\AbstractObject $upda
                     ]
-                ).'\') ORDER BY o_index LIMIT '. $updatedObject->getParent()->getChildAmount() .')
-                            SELECT @n := IF(@n = ? - 1,@n + 2,@n + 1) AS newIndex, o_id
+                ).'\') ORDER BY indexLIMIT '. $updatedObject->getParent()->getChildAmount() .')
```
there, a space after `index` is necessary.

fashxp commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1366,22 +1366,22 @@ protected function updateIndexesOfObjectSiblings(DataObject\AbstractObject $upda
+                        SELECT newIndex, id
                         FROM (
-                            With cte As (SELECT o_index, o_id FROM ' . $list->getDao()->getTableName() . ' WHERE o_parentId = ? AND o_id != ? AND o_type IN (\''.implode(
+                            With cte As (SELECT index, id FROM ' . $list->getDao()->getTableName() . ' WHERE parentId = ? AND id != ? AND type IN (\''.implode(
```
`index` needs to be quoted ... it is a reserved word in mysql

fashxp replied:
also consider for other appearances

fashxp commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1059,9 +1059,9 @@ public function deleteAction(Request $request): JsonResponse
 
             $list = new DataObject\Listing();
-            $list->setCondition('o_path LIKE ' . $list->quote($list->escapeLike($parentObject->getRealFullPath()) . '/%'));
+            $list->setCondition('path LIKE ' . $list->quote($list->escapeLike($parentObject->getRealFullPath()) . '/%'));
```
`path` is a reserved word and should be quoted

fashxp replied:
also consider for other appearances

mcop1 replied:
`path` isn´t a reserved word for column names, right? If we quote `path` anyway, then we should quote `type` too.

mcop1 replied:
Solution: quote both

fashxp commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1390,8 +1390,8 @@ protected function updateIndexesOfObjectSiblings(DataObject\AbstractObject $upda
-                'SELECT o_id, o_modificationDate, o_versionCount, o_key, o_index FROM objects'
-                ." WHERE o_parentId = ? AND o_id != ? AND o_type IN ('object', 'variant','folder') ORDER BY o_index ASC",
+                'SELECT id, modificationDate, versionCount, key, index FROM objects'
+                ." WHERE parentId = ? AND id != ? AND type IN ('object', 'variant','folder') ORDER BY indexASC",
```
space after `index` missing

fashxp commented on bundles/CoreBundle/src/Migrations/Version20221003115124.php:
```diff
@@ -0,0 +1,95 @@
+                    $this->addSql("ALTER TABLE $tableName CHANGE o_id id int(11) unsigned NOT NULL;");
+                }
+            }
+        }
```
all product index tables are missing ... which have additional columns like `o_virtualProductId`, `o_virtualProductActive`, `o_classId`, `o_parentId`, `o_type`

fashxp replied:
we also need to find a way how to "update" or "migrate" eventually existing elasticsearch indices ... or add upgrade notes, that they need to be rebuild...

fashxp replied:
we also need to update or regenerate all `object_*` views

mcop1 replied:
A hint for product index tables and elastic search indexes will be added to the upgrade notes.
`object_*` views will be regenerated by classes-rebuild in a separate migration -> https://github.com/pimcore/pimcore/issues/12859

fashxp commented on models/DataObject/AbstractObject.php:
```diff
@@ -235,36 +233,21 @@ abstract class AbstractObject extends Model\Element\AbstractElement
     protected function getBlockedVars(): array
     {
-        $blockedVars = ['o_hasChildren', 'o_versions', 'o_class', 'scheduledTasks', 'o_parent', 'parent', 'omitMandatoryCheck'];
+        $blockedVars = ['hasChildren', 'versions', 'class', 'scheduledTasks', 'parent', 'parent', 'omitMandatoryCheck'];
```
up to now, `hasChildren`, `siblings`, `hasSiblings`, `childrenSortBy`, `childrenSortOrder`, `versionCount`, `dirtyLanguages` and `dirtyFields` were not reserved ... that means, there might be classes with such fields. 
we need at least add that to the upgrade notes and require the devs to rename/migrate these fields before updating to Pimcore 11

mcop1 replied:
Added an entry to the upgrade notes

fashxp replied:
but we also need to add them here in the blockedVars list, right?

fashxp replied:
actually not here, but [here](https://github.com/pimcore/pimcore/blob/815ec901d89af3f57f458012993357f269c39eef/models/DataObject/ClassDefinition/Data.php#L77) and [here](https://github.com/pimcore/pimcore/blob/f25e567a626c7c6e52f53846d9ed9a5a9b374721/bundles/AdminBundle/public/js/pimcore/object/classes/data/data.js#L18)

mcop1 replied:
Added them to the forbidden names array

fashxp commented on models/DataObject/ClassDefinition/Data/AdvancedManyToManyRelation.php:
```diff
@@ -264,12 +264,12 @@ public function getDataForEditmode(mixed $data, DataObject\Concrete $object = nu
+                    $pathCol = 'path';
+                    $keyCol = 'key';
+                    $typeCol = 'type';
+                    $className = ', className className';
```
what is that?

mcop1 replied:
`className` was used as an alias in the following sql. This shouldn´t be necessary anymore. I also adapted the code a bit to eliminate the redundant variable assignments.

fashxp commented on models/Element/Service.php:
```diff
@@ -625,13 +611,13 @@ public static function minimizePropertiesForEditmode(array $props): array
             $allowedProperties = [
                 'key',
-                'o_key',
+                'key',
```
double entries...

fashxp commented on models/Element/Tag/Dao.php:
```diff
@@ -231,7 +231,7 @@ public function getElementsForTag(
             'document' => ['documents', 'id', 'type', '\Pimcore\Model\Document'],
             'asset' => ['assets', 'id', 'type', '\Pimcore\Model\Asset'],
-            'object' => ['objects', 'o_id', 'o_type', '\Pimcore\Model\DataObject\AbstractObject'],
+            'object' => ['objects', 'id', 'type', '\Pimcore\Model\DataObject\AbstractObject'],
```
we don't need that map anymore then, right?

fashxp replied:
at least `id` and `type`

mcop1 replied:
Removed `id` and `type` from the array and adapted the code.

fashxp commented:
Things to test
- [x] Test grid config with o_ prefixes
- [x] Check version compare
- [x] Check vouchers & onlineshop orders
- [ ] Check elasticsearch index
- [x] Check internal search
- [ ] Check saving of localized fields
- [x] Check tags
- [x] Check delete user

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Reviewer feedback

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Reviewer feedback

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

fashxp committed:
Update README.md

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Reviewer feedback

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
Added property names to forbidden names array

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

fashxp commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1059,9 +1059,9 @@ public function deleteAction(Request $request): JsonResponse
+            $list->setCondition('`path` LIKE ' . $list->quote($list->escapeLike($parentObject->getRealFullPath()) . '/%'));
             $list->setLimit((int)$request->get('amount'));
-            $list->setOrderKey('LENGTH(o_path)', false);
+            $list->setOrderKey('LENGTH(path)', false);
```
quote

fashxp commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1881,8 +1881,8 @@ public function copyInfoAction(Request $request): JsonResponse
-                $list->setCondition('o_path LIKE ' . $list->quote($list->escapeLike($object->getRealFullPath()) . '/%'));
-                $list->setOrderKey('LENGTH(o_path)', false);
+                $list->setCondition('`path` LIKE ' . $list->quote($list->escapeLike($object->getRealFullPath()) . '/%'));
+                $list->setOrderKey('LENGTH(path)', false);
```
quote

fashxp commented on bundles/AdminBundle/src/Controller/Admin/TagsController.php:
```diff
@@ -316,14 +316,14 @@ public function getBatchAssignmentJobsAction(Request $request, EventDispatcherIn
             $userIds[] = $this->getAdminUser()->getId();
             $condition .= ' AND (
-                (SELECT `view` FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') and LOCATE(CONCAT(o_path,o_key),cpath)=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
+                (SELECT `view` FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') and LOCATE(CONCAT(path,key),cpath)=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
```
quote

fashxp commented on bundles/AdminBundle/src/Controller/Admin/TagsController.php:
```diff
@@ -316,14 +316,14 @@ public function getBatchAssignmentJobsAction(Request $request, EventDispatcherIn
+                (SELECT `view` FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') and LOCATE(CONCAT(path,key),cpath)=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
                     OR
-                (SELECT `view` FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') and LOCATE(cpath,CONCAT(o_path,o_key))=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
+                (SELECT `view` FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') and LOCATE(cpath,CONCAT(path,key))=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
```
quote

fashxp commented on bundles/AdminBundle/src/Helper/GridHelperService.php:
```diff
@@ -310,16 +310,16 @@ public function getFilterCondition(string $filterJson, ClassDefinition $class, $
                         // system field
                         if ($filterField == 'fullpath') {
-                            $conditionPartsFilters[] = 'concat(o_path, o_key) ' . $operator . ' ' . $db->quote('%' . $filter['value'] . '%');
+                            $conditionPartsFilters[] = 'concat(path, key) ' . $operator . ' ' . $db->quote('%' . $filter['value'] . '%');
```
quote

fashxp commented on bundles/AdminBundle/src/Helper/GridHelperService.php:
```diff
@@ -310,16 +310,16 @@ public function getFilterCondition(string $filterJson, ClassDefinition $class, $
+                            $conditionPartsFilters[] = 'concat(path, key) ' . $operator . ' ' . $db->quote('%' . $filter['value'] . '%');
                         } elseif ($filterField == 'key') {
-                            $conditionPartsFilters[] = 'o_key ' . $operator . ' ' . $db->quote('%' . $filter['value'] . '%');
+                            $conditionPartsFilters[] = 'key ' . $operator . ' ' . $db->quote('%' . $filter['value'] . '%');
```
quote

fashxp commented on bundles/AdminBundle/src/Helper/GridHelperService.php:
```diff
@@ -528,7 +528,7 @@ public function prepareListingForGrid(array $requestParams, string $requestedLan
                     $orderKey = $colMappings[$orderKey];
                 } elseif ($orderKey === 'fullpath') {
-                    $orderKey = 'CAST(CONCAT(o_path, o_key) AS CHAR CHARACTER SET utf8) COLLATE utf8_general_ci';
+                    $orderKey = 'CAST(CONCAT(path, key) AS CHAR CHARACTER SET utf8) COLLATE utf8_general_ci';
```
quote

fashxp commented on bundles/AdminBundle/src/Helper/GridHelperService.php:
```diff
@@ -563,20 +563,20 @@ public function prepareListingForGrid(array $requestParams, string $requestedLan
             $userIds[] = $adminUser->getId();
             $conditionFilters[] = ' (
-                                                    (select list from users_workspaces_object where userId in (' . implode(',', $userIds) . ') and LOCATE(CONCAT(o_path,o_key),cpath)=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
+                                                    (select list from users_workspaces_object where userId in (' . implode(',', $userIds) . ') and LOCATE(CONCAT(path,key),cpath)=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
```
quote

fashxp commented on bundles/AdminBundle/src/Helper/GridHelperService.php:
```diff
@@ -563,20 +563,20 @@ public function prepareListingForGrid(array $requestParams, string $requestedLan
+                                                    (select list from users_workspaces_object where userId in (' . implode(',', $userIds) . ') and LOCATE(CONCAT(path,key),cpath)=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
                                                     OR
-                                                    (select list from users_workspaces_object where userId in (' . implode(',', $userIds) . ') and LOCATE(cpath,CONCAT(o_path,o_key))=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
+                                                    (select list from users_workspaces_object where userId in (' . implode(',', $userIds) . ') and LOCATE(cpath,CONCAT(path,key))=1  ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
```
quote

fashxp commented on models/DataObject/AbstractObject/Dao.php:
```diff
@@ -174,7 +174,7 @@ public function updateChildPaths(string $oldPath): ?array
             //update object child paths
             // we don't update the modification date here, as this can have side-effects when there's an unpublished version for an element
-            $this->db->executeQuery('update objects set o_path = replace(o_path,' . $this->db->quote($oldPath . '/') . ',' . $this->db->quote($this->model->getRealFullPath() . '/') . "), o_userModification = '" . $userId . "' where o_path like " . $this->db->quote(Helper::escapeLike($oldPath) . '/%') . ';');
+            $this->db->executeQuery('update objects set `path` = replace(path,' . $this->db->quote($oldPath . '/') . ',' . $this->db->quote($this->model->getRealFullPath() . '/') . "), userModification = '" . $userId . "' where `path` like " . $this->db->quote(Helper::escapeLike($oldPath) . '/%') . ';');
```
quote

fashxp commented on models/DataObject/AbstractObject/Dao.php:
```diff
@@ -206,7 +206,7 @@ public function getCurrentFullPath(): ?string
 
         try {
-            $path = $this->db->fetchOne('SELECT CONCAT(o_path,`o_key`) as o_path FROM objects WHERE o_id = ?', [$this->model->getId()]);
+            $path = $this->db->fetchOne('SELECT CONCAT(path,`key`) as `path` FROM objects WHERE id = ?', [$this->model->getId()]);
```
quote

fashxp commented on models/DataObject/AbstractObject/Dao.php:
```diff
@@ -419,7 +419,7 @@ public function getChildAmount(?array $objectTypes = [DataObject::OBJECT_TYPE_OB
     public function getTypeById(int $id): array
     {
-        $t = $this->db->fetchAssociative('SELECT o_type,o_className,o_classId FROM objects WHERE o_id = ?', [$id]);
+        $t = $this->db->fetchAssociative('SELECT type,className,classId FROM objects WHERE id = ?', [$id]);
```
quote

fashxp commented on models/DataObject/Data/ElementMetadata/Dao.php:
```diff
@@ -50,13 +50,13 @@ public function save(DataObject\Concrete $object, string $ownertype, string $own
     {
         if ($destinationType == 'object') {
-            $typeQuery = " AND (type = 'object' or type = '')";
+            $typeQuery = " AND (type = 'object' or `type` = '')";
```
quote

fashxp commented on models/DataObject/Data/ObjectMetadata/Dao.php:
```diff
@@ -58,9 +58,9 @@ protected function getTablename(DataObject\Concrete $object): string
     public function load(DataObject\Concrete $source, int $destinationId, string $fieldname, string $ownertype, string $ownername, string $position, int $index, string $destinationType = 'object'): ?DataObject\Data\ObjectMetadata
     {
-        $typeQuery = " AND (type = 'object' or type = '')";
+        $typeQuery = " AND (type = 'object' or `type` = '')";
```
quote

fashxp commented:
More findings: 

- [x] empty parent id in frontend 
![image](https://user-images.githubusercontent.com/8792145/207428077-09de47c5-95eb-4a33-a1ec-97fd36723afa.png)

- [x] event objects title is always set to `Event`
![image](https://user-images.githubusercontent.com/8792145/207434469-bfe58283-e327-4363-ac00-fd817de37a0c.png)

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

mcop1 committed:
[Data Objects] Remove o_ prefixed of system properties and database columns
More quotes and bugfix regarding the title of data object class

mcop1 commented:
> More findings:
> 
>     * [x]  empty parent id in frontend
>       ![image](https://user-images.githubusercontent.com/8792145/207428077-09de47c5-95eb-4a33-a1ec-97fd36723afa.png)
> 
>     * [x]  event objects title is always set to `Event`
>       ![image](https://user-images.githubusercontent.com/8792145/207434469-bfe58283-e327-4363-ac00-fd817de37a0c.png)

Problems fixed. Second one was a bit tricky, since the title from the data object class definition was set into the title of the actual object.

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

mcop1 committed:
Merge branch '11.x' into 7341-data-objects-remove-o_-prefixed-of-system-properties-and-database-columns

# Conflicts:
#	doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md
